File: TelerikUIExamples/CustomAnimationAreaChart.m
//
//  CustomAnimationAreaChart.m
//  TelerikUIExamples
//
//  Created by Svetlin Ralchev on 10/21/13.
//  Copyright (c) 2013 Telerik. All rights reserved.
//

#import "CustomAnimationAreaChart.h"
#import <TelerikUI/TelerikUI.h>

@interface CustomAnimationAreaChart ()

@end

@implementation CustomAnimationAreaChart
{
    TKChart *_chart;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    _chart = [[TKChart alloc] initWithFrame:[self exampleBounds]];
    _chart.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _chart.allowAnimations = YES;
    _chart.delegate = self;
    [self.view addSubview:_chart];
    
    NSMutableArray *points = [[NSMutableArray alloc] init];
    for (int i = 0; i<10; i++) {
        [points addObject:[[TKChartDataPoint alloc] initWithX:@(i) Y:@(arc4random() % 100)]];
    }
    
    TKChartAreaSeries *areaSeries = [[TKChartAreaSeries alloc] init];
    areaSeries.selectionMode = TKChartSeriesSelectionModeSeries;
    [_chart addSeries:areaSeries withItems:points];
}

#pragma mark TKChartDelegate

- (CAAnimation *)chart:(TKChart *)chart animationForSeries:(TKChartSeries *)series withState:(TKChartSeriesRenderState *)state inRect:(CGRect)rect
{
    CFTimeInterval duration = .5;
    NSMutableArray *animations = [[NSMutableArray alloc] init];
    
    for (int i = 0; i<state.points.count; i++) {
        NSString *keyPath = [NSString stringWithFormat:@"seriesRenderStates.%lu.points.%d.y", (unsigned long)series.index, i];
        TKChartVisualPoint *point = [state.points objectAtIndex:i];
        CGFloat oldY = rect.size.height;
        
        CGFloat half = oldY + (point.y - oldY)/2.;
        CAKeyframeAnimation *a = [CAKeyframeAnimation animationWithKeyPath:keyPath];
        a.keyTimes = @[ @0, @0, @1 ];
        a.values = @[ @(oldY), @(half), @(point.y) ];
        a.duration = duration;
        a.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
        [animations addObject:a];
    }
    
    CAAnimationGroup *group = [[CAAnimationGroup alloc] init];
    group.duration = duration;
    group.animations = animations;
    return group;
}

@end

File: TelerikUIExamplesInCSharp/Examples/Chart/Animations/CustomAnimationAreaChart.cs
using System;
using System.Collections.Generic;
using System.Drawing;

using Foundation;
using UIKit;
using CoreAnimation;
using CoreGraphics;

using TelerikUI;

namespace Examples
{
	[Register("CustomAnimationAreaChart")]
	public class CustomAnimationAreaChart: XamarinExampleViewController
	{
		ChartDelegate chartDelegate = new ChartDelegate ();

		public override void ViewDidLoad ()
		{
			base.ViewDidLoad ();

			TKChart chart = new TKChart (this.View.Bounds);
			chart.AutoresizingMask = UIViewAutoresizing.FlexibleWidth | UIViewAutoresizing.FlexibleHeight;
			chart.Delegate = chartDelegate;
			chart.AllowAnimations = true;
			this.View.AddSubview (chart);

			Random r = new Random ();
			List<TKChartDataPoint> list = new List<TKChartDataPoint> ();
			for (int i = 0; i < 10; i++) {
				list.Add (new TKChartDataPoint(new NSNumber(i), new NSNumber(r.Next() % 100)));
			}

			TKChartAreaSeries areaSeries = new TKChartAreaSeries (list.ToArray ());
			chart.AddSeries (areaSeries);
		}
			
		class ChartDelegate: TKChartDelegate
		{
			public override CAAnimation AnimationForSeries (TKChart chart, TKChartSeries series, TKChartSeriesRenderState state, CGRect rect)
			{
				double duration = 0.5;
				List<CAAnimation> animations = new List<CAAnimation> ();

				for (int i = 0; i<(int)state.Points.Count; i++) {
					string keyPath = string.Format ("seriesRenderStates.{0}.points.{1}.y", series.Index, i);
					TKChartVisualPoint point = (TKChartVisualPoint)state.Points.ObjectAtIndex((uint)i);
					double oldY = rect.Height;	
					double half = oldY + (point.Y - oldY)/2.0;
					CAKeyFrameAnimation a = (CAKeyFrameAnimation)CAKeyFrameAnimation.GetFromKeyPath(keyPath);
					a.KeyTimes = new NSNumber[] { new NSNumber (0), new NSNumber (0), new NSNumber (1) };
					a.Values = new NSObject[] { new NSNumber (oldY), new NSNumber (half), new NSNumber (point.Y) };
					a.Duration = duration;
					a.TimingFunction = CAMediaTimingFunction.FromName (CAMediaTimingFunction.EaseOut);
					animations.Add(a);
				}
	
				CAAnimationGroup group = new CAAnimationGroup ();

				group.Duration = duration;
				group.Animations = animations.ToArray();

				return group;
			}
		}
	}
}

